src/java.base/unix/classes/sun/nio/fs/UnixFileSystemProvider.java

```diff
@@ -348,8 +349,38 @@ public boolean isExecutable(Path path) {
         return access(file, X_OK) == 0;
     }
 
+    // if path is a symbolic link, then add its key to the set and
+    // continue recursively down any singly linked list of targets
+    private void getLinkKeys(UnixPath path, Set<UnixFileKey> keys)
+        throws IOException {
+        try {
+            UnixFileAttributes attrs = UnixFileAttributes.get(path, false);
+            if (attrs.isSymbolicLink()) {
+                if (!keys.add(attrs.fileKey()))
+                    return; // if key already present, then there is a loop so bail out
+                UnixPath target = UnixPath.toUnixPath(readSymbolicLink(path));
+                if (exists(target, LinkOption.NOFOLLOW_LINKS))
+                    getLinkKeys(target, keys);
+            }
+        } catch (UnixException x) {
+            x.rethrowAsIOException(path);
+        }
+    }
+
+    // retrieve the file keys of any symbolic links in this path
+    private Set<UnixFileKey> getLinkKeys(UnixPath path) throws IOException {
+        Set<UnixFileKey> keys = new HashSet<>();
+        UnixPath p = path;
+        while (p != null && exists(p, LinkOption.NOFOLLOW_LINKS)) {
+            getLinkKeys(p, keys);
+            p = p.getParent();
+        }
+        return keys;
+    }
```

Comment:
I'm a bit puzzled as to why it looks at each element in the path. It should be possible to just enumerate the files of the sym links that exist, something like this
```
        var fileKeys = new HashSet<UnixFileKey>();
        while (path != null) {
            UnixFileAttributes attrs = UnixFileAttributes.getIfExists(path, false);
            if (attrs != null && attrs.isSymbolicLink()) {
                fileKeys.add(attrs.fileKey());
                path = (UnixPath) readSymbolicLink(path);
            } else {
                path = null;
            }
        }
```